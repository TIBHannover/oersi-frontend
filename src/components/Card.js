import React, {useState} from "react"
import {useTranslation} from "react-i18next"
import {useNavigate} from "react-router-dom"
import PropTypes from "prop-types"
import {
  Button,
  Card as MuiCard,
  CardActions,
  CardContent,
  CardHeader,
  CardMedia,
  IconButton,
  Link,
  Typography,
  useTheme,
} from "@mui/material"
import LazyLoad from "react-lazyload"

import {getLicenseIcon, hasLicenseIcon} from "./CustomIcons"
import {
  getLicenseGroup,
  getSafeUrl,
  getThumbnailUrl,
  joinArrayField,
} from "../helpers/helpers"
import {OersiConfigContext} from "../helpers/use-context"

const Card = (props) => {
  const navigate = useNavigate()
  const theme = useTheme()
  const {t} = useTranslation(["translation", "language", "labelledConcept"])
  const oersiConfig = React.useContext(OersiConfigContext)
  const defaultImage = props.image
    ? props.image
    : process.env.PUBLIC_URL + "/help_outline.svg"
  const [thumbnailUrl, setThumbnailUrl] = useState(
    oersiConfig.FEATURES?.OERSI_THUMBNAILS
      ? getThumbnailUrl(props._id)
      : defaultImage
  )

  return (
    <React.Fragment>
      <MuiCard className="card-root" sx={{margin: theme.spacing(1.5)}}>
        <Link
          target="_blank"
          rel="noopener"
          href={getSafeUrl(props.id)}
          className="card-header-link"
          aria-label={props.name}
          underline="hover"
          color="textSecondary"
          sx={{fontWeight: theme.typography.fontWeightBold}}
        >
          <LazyLoad offset={100} once>
            <CardMedia
              className="card-media"
              image={thumbnailUrl}
              title={props.id}
              aria-label="resource image"
              sx={{height: 0, paddingTop: "56.25%"}}
            >
              {oersiConfig.FEATURES?.OERSI_THUMBNAILS && (
                <img
                  src={thumbnailUrl}
                  style={{display: "None"}}
                  onError={(e) => {
                    e.target.onerror = null
                    setThumbnailUrl(defaultImage)
                  }}
                  alt="fallback workaround"
                />
              )}
            </CardMedia>
          </LazyLoad>
          <CardHeader
            className="card-header-title"
            sx={{
              alignItems: "start",
              paddingTop: theme.spacing(1),
              paddingBottom: 0,
            }}
            title={
              <>
                <Typography
                  variant="subtitle2"
                  align="left"
                  component="div"
                  className={"card-header-title-creators-label card-text"}
                  sx={{
                    fontWeight: theme.typography.fontWeightBold,
                  }}
                >
                  {joinArrayField(props.creator, (item) => item.name) + " "}
                </Typography>
                <Typography
                  variant="h6"
                  align="left"
                  component="div"
                  className={"card-header-title-label card-text"}
                  color="primary"
                  sx={{
                    fontWeight: theme.typography.fontWeightBold,
                  }}
                >
                  {props.name}
                </Typography>
              </>
            }
          />
        </Link>
        <CardContent className="card-infos" sx={{paddingY: theme.spacing(1)}}>
          {getDescription()}
          {getCardInfoTextEntry(
            joinArrayField(
              props.about,
              (item) => item.id,
              (label) =>
                t("labelledConcept#" + label, {
                  keySeparator: false,
                  nsSeparator: "#",
                })
            )
          )}
          {getCardInfoTextEntry(
            joinArrayField(
              props.learningResourceType,
              (item) => item.id,
              (label) =>
                t("labelledConcept#" + label, {
                  keySeparator: false,
                  nsSeparator: "#",
                })
            )
          )}
        </CardContent>
        <CardActions
          className="card-actions"
          sx={{marginTop: "auto"}}
          disableSpacing
        >
          <div>{getLicense()}</div>
          <Button
            color="grey"
            className="button-details"
            onClick={() => navigate("/" + props._id)}
          >
            {t("LABEL.SHOW_DETAILS")}
          </Button>
        </CardActions>
      </MuiCard>
    </React.Fragment>
  )

  function getDescription() {
    let content = null
    if (props.description) {
      content = props.description
    } else if (props.keywords) {
      content = joinArrayField(props.keywords, (item) => item)
    }
    if (content) {
      return (
        <Typography
          variant="body1"
          className={"card-description card-text"}
          color="textPrimary"
          sx={{fontWeight: 500}}
        >
          {content}
        </Typography>
      )
    }
    return ""
  }

  function getLicense() {
    if (props.license && props.license.id) {
      const licenseGroup = getLicenseGroup(props.license)
      return !licenseGroup || hasLicenseIcon(licenseGroup.toLowerCase()) ? (
        <IconButton
          className="card-action-license"
          target="_blank"
          rel="noreferrer"
          href={getSafeUrl(props.license.id)}
          aria-label={licenseGroup}
          size="large"
        >
          {getLicenseIcon(licenseGroup.toLowerCase())}
        </IconButton>
      ) : (
        <Button
          color="grey"
          className="card-action-license"
          target="_blank"
          rel="noreferrer"
          href={getSafeUrl(props.license.id)}
          aria-label={licenseGroup}
        >
          {licenseGroup}
        </Button>
      )
    }
    return ""
  }
  function getCardInfoTextEntry(text, ariaLabel) {
    return text ? (
      <Typography
        variant="body1"
        aria-label={ariaLabel}
        className={"card-info card-text"}
        sx={{marginTop: theme.spacing(1)}}
        component="div"
      >
        {text}
      </Typography>
    ) : (
      ""
    )
  }
}

Card.propTypes = {
  props: PropTypes.object,
}

export default Card
