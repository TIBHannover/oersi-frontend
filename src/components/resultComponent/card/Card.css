.card {
  font-weight: 400;
}
.card-item-row {
  margin-top: 10px;
}
.card-item-card {
  -webkit-box-shadow: 9px 10px 20px -10px #516c74;
  box-shadow: 9px 10px 20px -10px #3f4a4d;
}
.card-img {
  width: 100% !important;
}

.stats {
  margin-left: 2%;
  font-size: 100%;
  float: left; /*Add float left*/
}

.author-text {
  margin-left: 30px;
  font-weight: 400;
  font-size: 14px;
}

.text-center {
  margin: 10px;
}

.card-footer {
  padding: 0;
  background-color: transparent;
  line-height: 30px;
}

.legend-span {
  white-space: unset;
  overflow: hidden;
  text-overflow: ellipsis;
}

.block-with-text {
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}

.block-with-text:active {
  display: block;
}

.licence {
  display: flex;
}

.licence-image {
  float: right;
  width: 80px;
  margin-left: 10px;
}
