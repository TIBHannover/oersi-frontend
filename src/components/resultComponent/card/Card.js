import React from "react"
import "./Card.css"
import moment from "moment"
import "moment/locale/de"
import {withTranslation} from "react-i18next"
import PropTypes from "prop-types"
import {makeStyles} from "@material-ui/core/styles"
import clsx from "clsx"
import Card from "@material-ui/core/Card"
import CardHeader from "@material-ui/core/CardHeader"
import CardMedia from "@material-ui/core/CardMedia"
import CardContent from "@material-ui/core/CardContent"
import CardActions from "@material-ui/core/CardActions"
import Collapse from "@material-ui/core/Collapse"
import IconButton from "@material-ui/core/IconButton"
import Typography from "@material-ui/core/Typography"
import ExpandMoreIcon from "@material-ui/icons/ExpandMore"
import Grid from "@material-ui/core/Grid"
import StorageIcon from "@material-ui/icons/Storage"
import InsertDriveFileIcon from "@material-ui/icons/InsertDriveFile"
import GTranslateIcon from "@material-ui/icons/GTranslate"
import Chip from "@material-ui/core/Chip"
import Link from "@material-ui/core/Link"
import Iso639Type from "iso-639-language"
import i18next from "i18next"
import {ConfigurationRunTime} from "../../../helpers/use-context"

const useStyles = makeStyles((theme) => ({
  expand: {
    transform: "rotate(0deg)",
    marginLeft: "auto",
    transition: theme.transitions.create("transform", {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: "rotate(180deg)",
  },
}))

const Cards = (props) => {
  const classes = useStyles()
  const [expanded, setExpanded] = React.useState(false)
  const context = React.useContext(ConfigurationRunTime)
  const iso639_1 = Iso639Type.getType(1)
  const handleExpandClick = () => {
    setExpanded(!expanded)
  }
  return (
    <React.Fragment>
      <Card className="card-card-root">
        <CardHeader
          className="card-card-header"
          title={
            <Link
              target="_blank"
              href={props.mainEntityOfPage[0].id}
              className="card-card-header-link"
            >
              {props.name}
            </Link>
          }
          subheader={formatDate(props.mainEntityOfPage[0].dateModified, "ll")}
        />
        <Grid container spacing={3} className="card-card-grid-container">
          <Grid item xs={12} sm={6}>
            {/* There is already an h1 in the page, let's not duplicate it. */}
            <Typography variant="body1" className="card-card-author" component="p">
              <b>{props.t("CARD.AUTHOR")}:</b> {props.creator[0].name}
            </Typography>
          </Grid>
          <Grid item xs={12} sm={6}>
            {props.license !== null && (
              <Link target="_blank" href={props.license} color="inherit">
                <Typography
                  variant="body1"
                  className="card-card-license"
                  component="p"
                >
                  <b>{props.t("CARD.LICENSE")}: </b>{" "}
                  <img
                    width="100px"
                    height="40"
                    src={
                      process.env.PUBLIC_URL +
                      "/licence/" +
                      licenseSplit(props.license).toLowerCase() +
                      ".svg"
                    }
                    alt={licenseSplit(props.license).toLowerCase()}
                  />
                </Typography>
              </Link>
            )}
          </Grid>
          {/* <Grid item xs={12} md={12} sm={12}> */}
          <Grid item xs={12} lg={6} md={12} sm={12}>
            <Link
              target="_blank"
              href={props.mainEntityOfPage[0].id}
              color="inherit"
            >
              <CardMedia
                className="card-card-media"
                image={props.image}
                title={props.image}
              />
            </Link>
          </Grid>
          <Grid item xs={12} md={12} lg={6} sm={12}>
            <CardContent className="card-card-content">
              <Typography variant="h6" className="card-card-typografi-h6">
                {props.description}
              </Typography>
            </CardContent>
          </Grid>
          {/* </Grid> */}
        </Grid>

        <CardActions disableSpacing>
          <div className="card-card-chip-root">
            <Chip
              icon={<InsertDriveFileIcon />}
              label={props.t("lrt#" + props.learningResourceType.id, {
                keySeparator: false,
                nsSeparator: "#",
              })}
              // onClick={handleClick}
              // onDelete={handleDelete}
            />
            <Chip
              icon={<GTranslateIcon />}
              label={
                props.inLanguage !== null &&
                iso639_1.getNameByCodeTranslate(
                  props.inLanguage.toString().toLowerCase(),
                  i18next.language
                ) !== ""
                  ? iso639_1.getNameByCodeTranslate(
                      props.inLanguage.toString().toLowerCase(),
                      i18next.language
                    )
                  : props.inLanguage
              }
              // onClick={handleClick}
              // onDelete={handleDelete}
            />
            {props.mainEntityOfPage
              .filter((e) => e.provider && e.provider.name)
              .map((item) => {
                return (
                  <Link
                    target="_blank"
                    href={item.id}
                    key={item.provider.name}
                    className="card-card-chip-root"
                  >
                    <Chip
                      icon={<StorageIcon />}
                      label={props.t("provider:" + item.provider.name, {
                        keySeparator: false,
                      })}
                    />
                  </Link>
                )
              })}
          </div>
          <IconButton
            className={clsx(classes.expand, {
              [classes.expandOpen]: expanded,
            })}
            onClick={handleExpandClick}
            aria-expanded={expanded}
            aria-label="show more"
          >
            <ExpandMoreIcon />
          </IconButton>
        </CardActions>
        <Collapse in={expanded} timeout="auto" unmountOnExit>
          <CardContent>
            <Typography className="card-card-typografi-h6" paragraph>
              {props.t("CARD.PARAGRAF_TEXT")}:
            </Typography>
            <Typography className="card-card-typografi-h6" paragraph>
              {props.description}
            </Typography>
          </CardContent>
        </Collapse>
      </Card>
    </React.Fragment>
  )

  /**
   * split the license and get last 2 chars
   * @param {string} license
   */
  function licenseSplit(license) {
    if (license !== null) return license.split("/").slice(-2)[0]
    else return ""
  }

  function formatDate(date, format) {
    if (date !== null) {
      moment.locale(context.LANGUAGE)
      return moment(date).format(format)
    } else {
      return ""
    }
  }
}

Card.propTypes = {
  props: PropTypes.object,
}

export default withTranslation(["translation", "provider", "lrt"])(Cards)
