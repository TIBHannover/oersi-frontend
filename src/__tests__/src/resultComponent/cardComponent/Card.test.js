import React from "react"
import {shallow} from "../../../../setupFiles"
import Card from "../../../../components/resultComponent/card/Card"

describe("Card ==> Test UI  ", () => {
  // registerConfiguration()
  const wrapperShadow = shallow(<Card />)
  it("Card : should render correctly", () => {
    expect(wrapperShadow).toMatchSnapshot()
  })
})
