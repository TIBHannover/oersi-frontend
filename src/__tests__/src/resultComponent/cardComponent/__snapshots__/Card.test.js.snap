// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent ==> Test UI   AppComponent : should render correctly 1`] = `ShallowWrapper {}`;

exports[`Card ==> Test UI   Card : should render correctly 1`] = `ShallowWrapper {}`;

exports[`CardCompnent ==> Test UI   CardCompnent : should render correctly 1`] = `ShallowWrapper {}`;
